run-make: arm command wrappers with drop bombs

jieyouxu opened the pull request:
This PR is one in a series of cleanups to run-make tests and the run-make-support library.

### Summary

It's easy to forget to actually executed constructed command wrappers, e.g. `rustc().input("foo.rs")` but forget the `run()`, so to help catch these mistakes, we arm command wrappers with drop bombs on construction to force them to be executed by test code.

This PR also removes the `Deref`/`DerefMut` impl for our custom `Command` which derefs to `std::process::Command` because it can cause issues when trying to use a custom command:

```rs
htmldocck().arg().run()
```

fails to compile because the `arg()` is resolved to `std::process::Command::arg`, which returns `&mut std::process::Command` that doesn't have a `run()` command.

This PR also:

- Removes `env_var` on the `impl_common_helper` macro that was wrongly named and is a footgun (no users).
- Bumps the run-make-support library to version `0.1.0`.
- Adds a changelog to the support library.

### Details

Especially for command wrappers like `Rustc`, it's very easy to build up
a command invocation but forget to actually execute it, e.g. by using
`run()`. This commit adds "drop bombs" to command wrappers, which are
armed on command wrapper construction, and only defused if the command
is executed (through `run`, `run_fail`).

If the test writer forgets to execute the command, the drop bomb will
"explode" and panic with an error message. This is so that tests don't
silently pass with constructed-but-not-executed command wrappers.

This PR is best reviewed commit-by-commit.

try-job: x86_64-msvc

Kobzol commented on tests/run-make/rustdoc-error-lines/rmake.rs:
```diff
@@ -4,9 +4,8 @@
-        String::from_utf8(rustdoc().input("input.rs").arg("--test").command_output().stdout)
-            .unwrap();
+    let output = rustdoc().input("input.rs").arg("--test").run_fail();
+    let output = String::from_utf8(output.stdout).unwrap();
```
This line will be all over the place. We could create our own `Output` wrapper with methods like `.stdout_utf8()` or something like that (doesn't need to be in this PR, ofc).

jieyouxu replied:
Good point, added it to my list of TODOs :3

Kobzol commented on src/tools/run-make-support/src/rustc.rs:
```diff
@@ -4,14 +4,17 @@ use std::io::Write;
 
 /// Construct a new `rustc` invocation.
+#[must_use]
 pub fn rustc() -> Rustc {
```
We could just mark `Rustc` with `#[must_use]`, to avoid forgetting to put these annotations on top of functions that create `Rustc`.

jieyouxu replied:
Good point, I moved `#[must_use]` annotations to the command wrapper structs themselves.

Kobzol commented on src/tools/run-make-support/src/drop_bomb/mod.rs:
```diff
@@ -0,0 +1,37 @@
+mod tests;
+
+#[derive(Debug)]
+pub(crate) struct DropBomb {
```
I experimented a bit more with `#[must_use]` and I think that it's not enough, so a drop bomb is a good idea :+1:

jieyouxu replied:
Yeah, the reason for using drop bombs was two-fold:

1. `#[must_use]` was not sufficient to guarantee that constructed command wrappers are *eventually* actually executed.
2. `#[must_use]` was noisy on intermediate builder methods (especially if a builder method wants to be conditionally called).

Kobzol commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3517,6 +3517,10 @@ impl<'test> TestCx<'test> {
 
+        // In test code we want to be very pedantic about values being silently discarded that are
+        // annotated with `#[must_use]`.
+        cmd.arg("-Dunused_must_use");
```
I feel like this should probably respect `deny-warnings = false` in `config.toml`, although not using a `#[must_use]` thing is probably serious enough bug that we should gate compilation for it even during development (vs just having unused variables).

jieyouxu replied:
I feel like this is different from "normal" code:

- I would not `#[deny(unused_must_use)]` for normal code (e.g. rustc), because there are often cases where it is noisy and it's better checked at CI to avoid blocking local dev.
- But for these rmake.rs tests, they are often short, and not using `#[must_use]` values are almost always bugs and are esepecially bad if that allows the test to silently pass. That is, imo, if the test author *really* wants to not use a `#[must_use]` value, they should probably `let _ = must_use()` or something better.

I'm not opposed to making this respect `deny-warnings = false` in `config.toml`, but if I'm running tests locally I personally *want* them to fail fast due to `-Dunused_must_use` instead of waiting for PR CI.

Kobzol replied:
Ok, agreed, let's keep it like this.

Kobzol reviewed (commented):
Looks good, left a few comments.

r? @Kobzol

jieyouxu commented:
@rustbot author

jieyouxu commented:
@rustbot ready

Kobzol commented on tests/run-make/same-lib-two-locations-no-panic/rmake.rs:
```diff
@@ -7,22 +7,22 @@
+    fs::create_dir(&tmp_dir_other).unwrap();
     rustc().input("foo.rs").crate_type("dylib").arg("-Cprefer-dynamic").run();
-    fs::rename(dynamic_lib("foo"), &tmp_dir_other);
+    fs::rename(dynamic_lib("foo"), tmp_dir_other.join(dynamic_lib_name("foo"))).unwrap();
```
This seems like it has changed the logic of the test. Was the test broken before?

jieyouxu replied:
I should've made this into a separate commit, but this test was subtly broken before in that the `fs::rename` actually failed (when I added an `unwrap()`) because the rename `to` is `tmp_dir_other` which is a directory

> On Unix, if `from` is a directory, `to` must also be an (empty) directory. If `from` is not a directory, `to` must also be not a directory. In contrast, on Windows, from can be anything, but to must not be a directory.

In particular, "If `from` is not a directory, `to` must also be not a directory." was violated (`dynamic_lib("foo")` is a file but `tmp_dir_other` is a directory) and produced a panic.

Kobzol replied:
Ah, right, I was wondering how it could before, it was swallowing an error. Ok, nice!

jieyouxu replied:
The test before would actually fail to test the behavior it was trying to test because of the failed rename (it was trying to test same lib in two locations don't cause a panic)

```rs
rustc().input("bar.rs").library_search_path(tmp_dir_other).run();
```

but this line would silently pass even if the compiler did panic because the rename failed.

bors commented:
:hourglass: Testing commit 7867fb75a377fb7105fc0679e50a05add842ba25 with merge 7df2b08465634878856453c9e1cf74deb58da486...
<!-- homu: {"type":"BuildStarted","head_sha":"7867fb75a377fb7105fc0679e50a05add842ba25","merge_sha":"7df2b08465634878856453c9e1cf74deb58da486"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9319882266/job/25655428566) [(plain)](https://github.com/rust-lang-ci/rust/commit/7df2b08465634878856453c9e1cf74deb58da486/checks/25655428566/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests\run-make\reset-codegen-1 stdout ----

error: rmake recipe failed to complete
status: exit code: 101
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\reset-codegen-1\\reset-codegen-1\\rmake.exe"
--- stderr -------------------------------
--- stderr -------------------------------
thread 'main' panicked at C:\a\rust\rust\tests\run-make\reset-codegen-1\rmake.rs:37:39:
called `Result::unwrap()` on an `Err` value: Os { code: 5, kind: PermissionDenied, message: "Access is denied." }
------------------------------------------


---- [run-make] tests\run-make\same-lib-two-locations-no-panic stdout ----
---- [run-make] tests\run-make\same-lib-two-locations-no-panic stdout ----

error: rmake recipe failed to complete
status: exit code: 101
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\same-lib-two-locations-no-panic\\same-lib-two-locations-no-panic\\rmake.exe"
--- stderr -------------------------------
thread 'main' panicked at C:\a\rust\rust\tests\run-make\same-lib-two-locations-no-panic\rmake.rs:21:35:
called `Result::unwrap()` on an `Err` value: Os { code: 5, kind: PermissionDenied, message: "Access is denied." }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
---
test result: FAILED. 254 passed; 2 failed; 97 ignored; 0 measured; 0 filtered out; finished in 78.21s

Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 1:23:05
make: *** [Makefile:101: ci-msvc-ps1] Error 1
  network time: Fri, 31 May 2024 16:11:02 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9319882266/job/25659404996)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9319882266/job/25659404996","builder_name":"checks-actions"} -->

jieyouxu commented:
Oh it's the weird permissions issues with removing files on msvc... need to debug why that happens
@rustbot author

jieyouxu commented:
@rustbot blocked waiting on https://github.com/rust-lang/rust/pull/125827 to fix the permission denied issue

jieyouxu commented:
@rustbot unblock

jieyouxu commented:
@bors rollup=iffy

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit ecb1c8b9a91b777c90e7fef04ad999774a256bdd with merge d0a77b478236d7b67226f5188e6ef800cc9a0bcf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ecb1c8b9a91b777c90e7fef04ad999774a256bdd","merge_sha":"d0a77b478236d7b67226f5188e6ef800cc9a0bcf"} -->

jieyouxu commented:
No bors, this is a *try* job, not a merge job!
@bors r- retry

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit ecb1c8b9a91b777c90e7fef04ad999774a256bdd with merge 8c58c104c366851cdcada848c435c81d36051110...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ecb1c8b9a91b777c90e7fef04ad999774a256bdd","merge_sha":"8c58c104c366851cdcada848c435c81d36051110"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9343303111/job/25714590163)
Build commit: 8c58c104c366851cdcada848c435c81d36051110 (`8c58c104c366851cdcada848c435c81d36051110`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9343303111/job/25714590163"},"merge_sha":"8c58c104c366851cdcada848c435c81d36051110"} -->

jieyouxu commented:
@rustbot ready

Kobzol commented:
> No bors, this is a try job, not a merge job!

Curious, why did you think that bors was attempting to merge this PR?

jieyouxu commented:
> > No bors, this is a try job, not a merge job!
> 
> Curious, why did you think that bors was attempting to merge this PR?

At the time it was shown as a pending PR on the merge queue, it wasn't a pending (try)

jieyouxu commented:
Fixed merge conflicts and fixed missing `.unwrap()` in `tests/run-make/bare-outfile`.

jieyouxu commented:
@bors rollup=never (basically will conflict/fail with any other PRs modifying run-make tests)

jieyouxu commented:
I think I messed up a rebase and some of these diffs look really wrong. Marking as draft for now.

bors commented:
:umbrella: The latest upstream changes (presumably #126104) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kobzol commented:
I'd suggest rebasing the first two commits on `master`, and then fix up any new unused_mut problems in a new commit, to avoid needless conflicts.

jieyouxu commented:
@rustbot author

jieyouxu committed:
run-make-support: bump version

jieyouxu committed:
run-make-support: add changelog

jieyouxu committed:
run-make-support: add drop bomb module

jieyouxu committed:
run-make-support: remove `env_var`

This is incorrectly named (it's actually `env_clear`), and is itself
a gigantic footgun: removing `TMPDIR` on Unix and `TMP`/`TEMP` on
Windows basically wrecks anything that relies on `std::env::temp_dir`
from functioning correctly. For example, this includes rustc's codegen.

jieyouxu committed:
run-make-support: arm command with drop bombs

- Update all command wrappers and command construction helpers with
  `#[track_caller]` where suitable to help the drop bomb panic message.
- Remove `Deref`/`DerefMut` for `Command` because it was causing issues
  with resolving to `std::process::Command` in a method call chain.

jieyouxu committed:
tests/run-make: update tests to use new API

jieyouxu commented:
### Changes since last reviewed

- Removed the `#[must_use]` changes to be done in a follow-up PR, because I wanted to get the more basic drop bomb changes and start a changelog.
- Removed `Deref`/`DerefMut` impl on the custom command (see PR description).
- Started a changelog.
- Updated tests.

PR description has been updated accordingly.

jieyouxu commented:
@rustbot ready

jieyouxu commented:
@rustbot rollup=never (likely conflicts with other rmake.rs PRs)

jieyouxu commented:
@bors rollup=never (likely conflicts with other rmake.rs PRs)

Kobzol commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -1,55 +1,125 @@
+/// sense of confidence.
+///
+/// [`run`]: Self::run
+/// [`run_fail`]: Self::run
```
`Self::run` -> `Self::run_fail`

jieyouxu replied:
Good catch!

Kobzol commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -1,55 +1,125 @@
-        let caller_location = std::panic::Location::caller();
-        let caller_line_number = caller_location.line();
-
+        self.drop_bomb.defuse();
```
nit: We could probably just defuse in `command_output`.

jieyouxu replied:
Changed to defuse in `command_output`.

Kobzol commented on src/tools/run-make-support/src/rustdoc.rs:
```diff
@@ -30,12 +32,14 @@ fn setup_common() -> Command {
 
 impl Rustdoc {
     /// Construct a bare `rustdoc` invocation.
+    #[track_caller]
```
Does this `#[track_caller]` really do anything? Is it required if you called e.g. `bare()` and then immediately dropped the result, to get a better error from the drop bomb?

jieyouxu replied:
Oh this is because I forgot a `#[track_caller]` on `bare_rustdoc()`. Oops, good catch.

Kobzol replied:
To be honest, I meant it in a completely different way :laughing: It seemed to me that most of the added `#[track_caller]`s in this PR don't do anything, I just picked this up at random when I sent the comment. That's what I was asking about.

jieyouxu replied:
Let me go double check if `#[track_caller]` works how I think it works

Kobzol replied:
E.g. [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9b6de795be0dcb27e7a6ab8c787c47c4), commenting out the track caller doesn't seem to change anything.

jieyouxu replied:
Yes: the answer is that this `#[track_caller]` is actually load-bearing for the panic message's line number:

Test rmake.rs:

```
use run_make_support::bare_rustdoc;

fn main() {
    bare_rustdoc();
}
```

produces

```
thread 'main' panicked at src/tools/run-make-support/src/drop_bomb/mod.rs:43:13:
command constructed but not executed at line 4: `/home/gh-jieyouxu/rust/build/aarch64-unknown-linux-gnu/stage1/bin/rustdoc`
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

without the `#[track_caller]` on the `bare_rustdoc` and transitively `Rustdoc::bare`, the line number is for rustdoc.rs inside the support library

```
thread 'main' panicked at src/tools/run-make-support/src/drop_bomb/mod.rs:43:13:
command constructed but not executed at line 10: `/home/gh-jieyouxu/rust/build/aarch64-unknown-linux-gnu/stage1/bin/rustdoc`
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

jieyouxu replied:
My understanding is that `#[track_caller]` is a *hint* for when `Location::caller()` is called to traverse up the callstack until an item body (or whatever the correct term is) who is not annotated with `#[track_caller]` is found.

Kobzol replied:
Ok, let's keep it then, thanks :)

Kobzol reviewed (commented):
Left a few nits, otherwise LGTM.

Kobzol commented:
Feel free to r=me once CI is green :)

jieyouxu commented:
@bors r=@Kobzol

bors commented:
:pushpin: Commit c1c06d670f29e6db8a200fde183dd620c528bf2b has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol c1c06d670f29e6db8a200fde183dd620c528bf2b -->
<!-- homu: {"type":"Approved","sha":"c1c06d670f29e6db8a200fde183dd620c528bf2b","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c1c06d670f29e6db8a200fde183dd620c528bf2b with merge 8d4c7487ca3c0777f6e07cf95d609943976dcfe5...
<!-- homu: {"type":"BuildStarted","head_sha":"c1c06d670f29e6db8a200fde183dd620c528bf2b","merge_sha":"8d4c7487ca3c0777f6e07cf95d609943976dcfe5"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9457296436/job/26050841480) [(plain)](https://github.com/rust-lang-ci/rust/commit/8d4c7487ca3c0777f6e07cf95d609943976dcfe5/checks/26050841480/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting run_make_support v0.1.0 (/checkout/src/tools/run-make-support)
error: unresolved link to `Commmand`
  --> src/tools/run-make-support/src/command.rs:17:9
   |
17 | /// a [`Commmand`] is constructed but never executed, the drop bomb will explode and cause the test
   |         ^^^^^^^^ no item named `Commmand` in scope
   = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9457296436/job/26052413544)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9457296436/job/26052413544","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors r-

jieyouxu commented:
Fixed the broken intra-doc link.

@bors r=@Kobzol rollup=never p=1 (there are a couple of PRs blocked on this transitively)

bors commented:
:pushpin: Commit 5ec3eef9e70551a7eef059cfc9d039f053805cab has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 5ec3eef9e70551a7eef059cfc9d039f053805cab -->
<!-- homu: {"type":"Approved","sha":"5ec3eef9e70551a7eef059cfc9d039f053805cab","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9464724264/job/26078780191)
Approved by: Kobzol
Pushing 20ba13c38e5b8451b7b09e4009efc9010bfa3f2c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Kobzol","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9464724264/job/26078780191"},"merge_sha":"20ba13c38e5b8451b7b09e4009efc9010bfa3f2c"} -->

rust-timer commented:
Finished benchmarking commit (20ba13c38e5b8451b7b09e4009efc9010bfa3f2c): [comparison URL](https://perf.rust-lang.org/compare.html?start=6a207f4ff28b76147514cfd6acea86db3d148e42&end=20ba13c38e5b8451b7b09e4009efc9010bfa3f2c&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=6a207f4ff28b76147514cfd6acea86db3d148e42&end=20ba13c38e5b8451b7b09e4009efc9010bfa3f2c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.1% | [2.1%, 2.1%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6a207f4ff28b76147514cfd6acea86db3d148e42&end=20ba13c38e5b8451b7b09e4009efc9010bfa3f2c&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=6a207f4ff28b76147514cfd6acea86db3d148e42&end=20ba13c38e5b8451b7b09e4009efc9010bfa3f2c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=6a207f4ff28b76147514cfd6acea86db3d148e42&end=20ba13c38e5b8451b7b09e4009efc9010bfa3f2c&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 678.233s -> 677.565s (-0.10%)
**Artifact size**: 320.05 MiB -> 319.41 MiB (-0.20%)
